src/main/scala/mesosphere/marathon/core/volume/Context.scala

```diff
@@ -0,0 +1,58 @@
+package mesosphere.marathon.core.volume
+
+import mesosphere.marathon.state.Volume
+import org.apache.mesos.Protos.{ CommandInfo, ContainerInfo }
+
+/**
+  * BuilderContext captures additional metadata required for decorating mesos Task protobufs with
+  * metadata pertaining to volumes
+  */
+sealed trait BuilderContext
+
+/**
+  * ContainerContext captures the builder that generates a Task's ContainerInfo
+  */
+final case class ContainerContext(ci: ContainerInfo.Builder) extends BuilderContext
+/**
+  * CommandContext captures the builder that generates a Task's CommandInfo
+  */
+final case class CommandContext(ct: ContainerInfo.Type, ci: CommandInfo.Builder) extends BuilderContext
+
+/**
+  * ContextUpdate implementations decorate Context with additional metadata for a given Volume.
+  */
+trait ContextUpdate {
+  /**
+    * Generates an updated BuilderContext for the given input context `c` and volume `v`.
+    * @param c is the initial input BuilderContext
+    * @param v is the volume that provides metadata with which the initial context is decorated
+    * @return a BuilderContext decorated with metadata from volume `v`; None if the context is not
+    *  supported by the storage provider associated with the given volume.
+    */
+  protected def updated[C <: BuilderContext](c: C, v: Volume): Option[C] = None
```

Comment:
Why this complicated signature? The `Option[C]` type is actually not used below. Just make it `C => C` with the identify if nothing is done.